lib/ddtrace/pin.rb

```diff
@@ -1,36 +1,53 @@
-# typed: true
-require 'ddtrace/utils/only_once'
+# typed: false
 
 module Datadog
-  # A {Datadog::Pin} (a.k.a Patch INfo) is a small class which is used to
-  # set tracing metadata on a particular traced object.
+  # A {Datadog::Pin} sets metadata on a particular object.
   # This is useful if you wanted to, say, trace two different
   # database clusters.
   class Pin
+    SETTER_METHOD = '='.freeze
+    OPTIONS_SEPARATOR = ', '.freeze
+
     def self.get_from(obj)
       return nil unless obj.respond_to? :datadog_pin
 
       obj.datadog_pin
     end
 
-    attr_accessor :app, :app_type, :config, :name, :service_name, :tags, :writer
+    def self.set_on(obj, **options)
+      if (pin = get_from(obj))
+        options.each { |k, v| pin[k] = v }
+      else
+        pin = new(**options)
+        pin.onto(obj)
+      end
+
+      pin
+    end
+
+    attr_accessor :options
+
+    def initialize(**options)
+      @options = options
+    end
+
+    def [](name)
+      @options[name]
+    end
+
+    def []=(name, value)
+      @options[name] = value
+    end
 
-    # TODO: remove aliases and leave `service` when we rename Datadog::Contrib::Configuration::Settings#service_name
-    # to Datadog::Contrib::Configuration::Settings#service
-    alias service= service_name=
-    alias service service_name
+    def method_missing(method_name, *args, &block)
+      return send(:[]=, method_name.to_s[0...-1].to_sym, *args) if method_name.to_s[-1] == SETTER_METHOD.freeze
+      return send(:[], method_name) if options.key?(method_name)
```

Comment:
If this is being kept for compatibility, would it make sense to mark it as deprecated?

Reply:
Could you add documentation to this `#method_missing` method, so we understand why it exists at a later inspection time.

Reply:
I removed this `method_missing` behavior... didn't like it. There's a replacement feature for gRPC support in `datadog_interceptor.rb` now.